Client/Frontend/Browser/FaviconManager.swift

```diff
@@ -31,14 +36,81 @@ class FaviconManager : TabHelper {
     func scriptMessageHandlerName() -> String? {
         return "faviconsMessageHandler"
     }
+    
+    internal class FaviconError: MaybeErrorType {
+        internal var description: String {
+            return "No Image Loaded"
+        }
+    }
+    
+    private func loadFavicons(tab: Tab, profile: Profile, favicons: [Favicon]) -> Deferred<Maybe<[Favicon]>> {
+        let deferred = Deferred<Maybe<[Favicon]>>()
+        var deferreds: [() -> Deferred<Maybe<Favicon>>]
+        deferreds = favicons.map { favicon in
+            return { () -> Deferred<Maybe<Favicon>> in
+                if let url = NSURL(string: favicon.url) {
+                    return self.getFavicon(tab, iconUrl: url, icon: favicon, profile: profile)
+                }
+                else {
+                    return deferMaybe(FaviconError())
+                }
+            }
+        }
+        
+        accumulate(deferreds) >>== { favicons in
+                deferred.fill(Maybe(success: favicons))
+        }
+        
+        return deferred
+    }
+    
+    func getFavicon(tab: Tab, iconUrl: NSURL, icon: Favicon, profile: Profile) -> Deferred<Maybe<Favicon>> {
+        let deferred = Deferred<Maybe<Favicon>>()
+        let manager = SDWebImageManager.sharedManager()
+        let options = tab.isPrivate ?
+            [SDWebImageOptions.LowPriority, SDWebImageOptions.CacheMemoryOnly] : [SDWebImageOptions.LowPriority]
+        
+        if let currentURL = tab.url,
+            let url = tab.url?.absoluteString {
+            let site = Site(url: url, title: "")
+            manager.downloadImageWithURL(iconUrl,
+                                         options: SDWebImageOptions(options),
+                                         progress: nil,
+                                         completed: { (img, err, cacheType, success, url) -> Void in
+                                            let fav = Favicon(url: url.absoluteString,
+                                                date: NSDate(),
+                                                type: icon.type)
+                                            
+                                            if let img = img {
```

Comment:
Thank you so much for helping me with this deferred stuff! It can be quite confusing at times.